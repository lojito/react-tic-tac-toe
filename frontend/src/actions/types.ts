import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

export type CategoryAction = {
  type: "CHANGE_CATEGORY";
  category: Categories;
};

export type DisabledAction = {
  type: "CHANGE_DISABLED";
  disabled: boolean;
};

export type FirstAction = {
  type: "CHANGE_FIRST";
  first: FirstPlayer;
};

export type UserImageAction = {
  type: "CHANGE_USER_IMAGE";
  userImage: ImageName;
};

export type ComputerImageAction = {
  type: "CHANGE_COMPUTER_IMAGE";
  computerImage: ImageName;
};

export type PlayingLevelAction = {
  type: "CHANGE_PLAYING_LEVEL";
  level: PlayingLevel;
};

export type GameOverAction = {
  type: "CHANGE_GAME_OVER";
  over: boolean;
};

export type ResultAction = {
  type: "CHANGE_GAME_RESULT";
  result: GameResult;
};

export type WinnersSquaresAction = {
  type: "CHANGE_WINNERS_SQUARES";
  winners: SquareLocation[];
};

export type ActionUnknow = {
  type: "UNKNOWN";
};

export type Action =
  | CategoryAction
  | DisabledAction
  | FirstAction
  | UserImageAction
  | ComputerImageAction
  | PlayingLevelAction
  | GameOverAction
  | ResultAction
  | WinnersSquaresAction
  | ActionUnknow;
