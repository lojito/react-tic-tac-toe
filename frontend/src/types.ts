export type CategoryId = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export type CategoryHavanaLandmarks = {
  id: 0;
  name: "CATEGORY_HAVANA_LANDMARKS";
  folder: "habana";
};

export type CategoryMontrealLandmarks = {
  id: 1;
  name: "CATEGORY_MONTREAL_LANDMARKS";
  folder: "montreal";
};

export type CategoryVancouverLandmarks = {
  id: 2;
  name: "CATEGORY_VANCOUVER_LANDMARKS";
  folder: "vancouver";
};

export type CategorySpainLandmarks = {
  id: 3;
  name: "CATEGORY_SPAIN_LANDMARKS";
  folder: "spain";
};

export type CategoryGermanyLandmarks = {
  id: 4;
  name: "CATEGORY_GERMANY_LANDMARKS";
  folder: "germany";
};

export type CategorySoccerPlayers = {
  id: 5;
  name: "CATEGORY_SOCCER_PLAYERS";
  folder: "soccer";
};

export type CategoryFruitsAndVegetables = {
  id: 6;
  name: "CATEGORY_FRUITS_AND_VEGETABLES";
  folder: "fruits";
};

export type CategoryAnimals = {
  id: 7;
  name: "CATEGORY_ANIMALS";
  folder: "animals";
};

export type CategoryPuppies = {
  id: 8;
  name: "CATEGORY_PUPPIES";
  folder: "puppies";
};

export type CategorySeinfeld = {
  id: 9;
  name: "CATEGORY_SEINFELD";
  folder: "seinfeld";
};

export type Categories =
  | CategoryAnimals
  | CategoryFruitsAndVegetables
  | CategoryGermanyLandmarks
  | CategoryHavanaLandmarks
  | CategoryMontrealLandmarks
  | CategoryPuppies
  | CategorySeinfeld
  | CategorySoccerPlayers
  | CategorySpainLandmarks
  | CategoryVancouverLandmarks;

export type ImageFolder =
  | "habana"
  | "montreal"
  | "vancouver"
  | "spain"
  | "germany"
  | "soccer"
  | "fruits"
  | "animals"
  | "puppies"
  | "seinfeld";

export type ImageName =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19;

export type FirstPlayer = Player.User | Player.Computer;

export enum PlayingLevel {
  Easy,
  Smart,
}

export enum Player {
  User,
  Computer,
  Nobody,
}

export type GameBoard = [
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player
];

export type SquareBoard = {
  disabled: boolean;
  id: string;
  imagePath: string;
  className: "win" | "";
};

export enum SquareLocation {
  TopLeft,
  TopCenter,
  TopRight,
  MiddleLeft,
  MiddleCenter,
  MiddleRight,
  BottomLeft,
  BottomCenter,
  BottomRight,
}

export enum GameResult {
  Draw,
  Lost,
  Started,
  Won,
}

export type DBGame = {
  _id?: string;
  board: GameBoard;
  categoryId: CategoryId;
  first: FirstPlayer;
  userImage: ImageName;
  computerImage: ImageName;
  level: PlayingLevel;
  result: GameResult;
  winners: SquareLocation[];
  user?: string;
  createdAt?: string;
  updatedAt?: string;
  __v?: number;
};
