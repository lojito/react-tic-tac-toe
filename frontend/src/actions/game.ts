import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

import {
  CategoryAction,
  ComputerImageAction,
  DisabledAction,
  FirstAction,
  GameOverAction,
  PlayingLevelAction,
  ResultAction,
  UserImageAction,
  WinnersSquaresAction,
} from "./types";

export const CHANGE_CATEGORY = "CHANGE_CATEGORY";
export const CHANGE_DISABLED = "CHANGE_DISABLED";
export const CHANGE_FIRST = "CHANGE_FIRST";
export const CHANGE_IMAGES = "CHANGE_IMAGES";
export const CHANGE_USER_IMAGE = "CHANGE_USER_IMAGE";
export const CHANGE_COMPUTER_IMAGE = "CHANGE_COMPUTER_IMAGE";
export const CHANGE_PLAYING_LEVEL = "CHANGE_PLAYING_LEVEL";
export const CHANGE_GAME_OVER = "CHANGE_GAME_OVER";
export const CHANGE_GAME_RESULT = "CHANGE_GAME_RESULT";
export const CHANGE_WINNERS_SQUARES = "CHANGE_WINNERS_SQUARES";

export const changeCategory = (category: Categories): CategoryAction => {
  return {
    type: CHANGE_CATEGORY,
    category,
  };
};

export const changeDisabled = (disabled: boolean): DisabledAction => {
  return {
    type: CHANGE_DISABLED,
    disabled,
  };
};

export const changeFirst = (first: FirstPlayer): FirstAction => {
  return {
    type: CHANGE_FIRST,
    first,
  };
};

export const changeUserImage = (userImage: ImageName): UserImageAction => {
  return {
    type: CHANGE_USER_IMAGE,
    userImage,
  };
};

export const changeComputerImage = (
  computerImage: ImageName
): ComputerImageAction => {
  return {
    type: CHANGE_COMPUTER_IMAGE,
    computerImage,
  };
};

export const changeLevel = (level: PlayingLevel): PlayingLevelAction => {
  return {
    type: CHANGE_PLAYING_LEVEL,
    level,
  };
};

export const changeOver = (over: boolean): GameOverAction => {
  return {
    type: CHANGE_GAME_OVER,
    over,
  };
};

export const changeResult = (result: GameResult): ResultAction => {
  return {
    type: CHANGE_GAME_RESULT,
    result,
  };
};

export const changeWinners = (
  winners: SquareLocation[]
): WinnersSquaresAction => {
  return {
    type: CHANGE_WINNERS_SQUARES,
    winners,
  };
};
