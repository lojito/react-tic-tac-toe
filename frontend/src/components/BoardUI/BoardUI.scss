@media screen and (max-width: 480px) {
  .board {
    width: 95%;
    margin: 0 auto;
  }
}
