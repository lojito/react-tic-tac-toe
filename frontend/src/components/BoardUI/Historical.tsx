import React, { FC, memo } from "react";
import {
  GameBoard,
  ImageFolder,
  ImageName,
  Player,
  SquareLocation,
} from "../../types";
import Square from "../Square/Historical";
import "./BoardUI.scss";

interface Props {
  folder: ImageFolder;
  computerImage: ImageName;
  userImage: ImageName;
  board: GameBoard;
  winners: SquareLocation[];
}

const Historical: FC<Props> = ({
  folder,
  userImage,
  computerImage,
  board,
  winners,
}) => {
  const boardJSX = board.map((player: Player, square: SquareLocation) => {
    const imagePath =
      player === Player.Computer
        ? `${folder}/${computerImage}`
        : player === Player.User
        ? `${folder}/${userImage}`
        : "default";
    const className = winners.indexOf(square) !== -1 ? "win" : "";

    return (
      <Square
        disabled
        id={square + ""}
        imagePath={imagePath}
        key={`square-${square}`}
        className={className}
      />
    );
  });

  return <div className="board">{boardJSX}</div>;
};

export default memo(Historical);
