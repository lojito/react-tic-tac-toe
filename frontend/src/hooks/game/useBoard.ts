import { useCallback, useEffect, useRef, useState } from "react";
import Board from "../../classes/Board";
import Computer from "../../classes/Computer";
import {
  Categories,
  DBGame,
  FirstPlayer,
  GameResult,
  ImageName,
  Player,
  PlayingLevel,
  SquareLocation,
} from "../../types";

interface Props {
  disabled: boolean;
  first: FirstPlayer;
  level: PlayingLevel;
  userImage: ImageName;
  computerImage: ImageName;
  category: Categories;
}

function useBoard(
  {
    disabled,
    first,
    level,
    userImage,
    computerImage,
    category: { id: categoryId },
  }: Props,
  onGameOver: (game: DBGame) => void
) {
  const board = useRef(Board);
  const computer = useRef(Computer);
  const gameOver = useRef(false);
  const player = useRef(Player.Nobody);
  const result = useRef(GameResult.Started);
  const winners = useRef<SquareLocation[]>([]);
  board.current.first = first;
  board.current.level = level;
  const [gameBoard, setGameBoard] = useState(board.current.getBoard());

  const playGame = useCallback((square?: SquareLocation) => {
    if (player.current === Player.Computer) {
      computer.current.play();
    } else {
      board.current.place(Player.User, square!);
    }
    setGameBoard(board.current.getBoard());

    winners.current = board.current.isAWinner(player.current);
    if (winners.current.length) {
      gameOver.current = true;
      result.current =
        player.current === Player.Computer ? GameResult.Lost : GameResult.Won;
    } else if (board.current.isFull) {
      gameOver.current = true;
      result.current = GameResult.Draw;
    } else {
      player.current =
        player.current === Player.Computer ? Player.User : Player.Computer;
      if (player.current === Player.Computer) {
        playGame();
      }
    }
  }, []);

  const resetGame = useCallback(() => {
    board.current.reset();
    gameOver.current = false;
    player.current = Player.Nobody;
    result.current = GameResult.Started;
    winners.current = [];
    setGameBoard(board.current.getBoard());
  }, []);

  useEffect(() => {
    if (!disabled && !gameOver.current) {
      resetGame();
    }
  }, [disabled, resetGame]);

  useEffect(() => {
    if (disabled === false) {
      if (first === Player.Computer) {
        player.current = Player.Computer;
      } else {
        player.current = Player.User;
      }
    } else {
      if (player.current === Player.Computer) {
        playGame();
      }
    }
  }, [disabled, first, playGame]);

  useEffect(() => {
    if (gameOver.current) {
      onGameOver({
        board: gameBoard,
        categoryId,
        first,
        userImage: userImage,
        computerImage: computerImage,
        level,
        result: result.current,
        winners: winners.current,
      });
      gameOver.current = false;
    }
  });

  return {
    playGame,
    gameBoard,
  };
}

export default useBoard;
