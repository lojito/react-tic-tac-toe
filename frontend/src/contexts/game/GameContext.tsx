import React, { createContext } from "react";
import { Action } from "../../actions/types";
import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../../types";

export type State = {
  category: Categories;
  userImage: ImageName;
  computerImage: ImageName;
  first: FirstPlayer;
  level: PlayingLevel;
  disabled: boolean;
  over: boolean;
  result: GameResult;
  winners: SquareLocation[];
};

export const GameContext = createContext<{
  game: State;
  dispatch: React.Dispatch<Action>;
}>({ game: {} as State, dispatch: {} as React.Dispatch<Action> });
