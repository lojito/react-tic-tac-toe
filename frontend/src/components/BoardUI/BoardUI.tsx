import React, { FC, memo, useContext } from "react";
import { GameContext } from "../../contexts/game/GameContext";
import useBoard from "../../hooks/game/useBoard";
import {
  DBGame,
  GameBoard,
  ImageFolder,
  ImageName,
  Player,
  SquareLocation,
} from "../../types";
import Square from "../Square/Square";
import "./BoardUI.scss";

interface MemoizedBoardProps {
  playGame: (id?: SquareLocation | undefined) => void;
  disabled: boolean;
  board: GameBoard;
  folder: ImageFolder;
  userImage: ImageName;
  computerImage: ImageName;
  over: boolean;
  winners: SquareLocation[];
}

const MemoizedBoard: FC<MemoizedBoardProps> = memo(
  ({
    disabled,
    board,
    folder,
    userImage,
    computerImage,
    over,
    winners,
    playGame,
  }) => {
    const squares = board.map((player: Player, square: SquareLocation) => {
      const imagePath =
        player === Player.Computer
          ? `${folder}/${computerImage}`
          : player === Player.User
          ? `${folder}/${userImage}`
          : "default";
      const isDisabled = !disabled || over || player !== Player.Nobody;
      const className = over
        ? winners.indexOf(square) !== -1
          ? "win"
          : ""
        : "";

      return (
        <Square
          disabled={isDisabled}
          id={square + ""}
          imagePath={imagePath}
          key={`square-${square}`}
          className={className}
        />
      );
    });

    return (
      <div
        className="board"
        onClick={(e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
          playGame(+(e.target as HTMLDivElement).id);
        }}
        data-testid="board"
      >
        {squares}
      </div>
    );
  }
);

interface Props {
  onGameOver: (game: DBGame) => void;
}

const BoardUI: FC<Props> = ({ onGameOver }) => {
  const { game } = useContext(GameContext);
  const { category, disabled, userImage, computerImage, over, winners } = game;
  const { folder } = category;

  const { playGame, gameBoard: board } = useBoard(game, onGameOver);

  return (
    <MemoizedBoard
      disabled={disabled}
      board={board}
      folder={folder}
      userImage={userImage}
      computerImage={computerImage}
      over={over}
      winners={winners}
      playGame={playGame}
    />
  );
};

export default BoardUI;
