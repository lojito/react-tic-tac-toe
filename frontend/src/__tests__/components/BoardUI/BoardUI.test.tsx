import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import BoardUI from "../../../components/BoardUI/BoardUI";
import { GameContext, State } from "../../../contexts/game/GameContext";
import useBoard from "../../../hooks/game/useBoard";
import { GameBoard, GameResult, Player, PlayingLevel } from "../../../types";

jest.mock(
  "../../../components/Square/Square",
  () =>
    ({
      disabled,
      id,
      imagePath,
      className,
    }: {
      disabled: boolean;
      id: string;
      imagePath: string;
      className: string;
    }) => {
      const path = `url(${process.env.PUBLIC_URL}/images`;
      const url =
        imagePath === "default"
          ? `${path}/default.jpg)`
          : `${path}/${imagePath}.jpg)`;

      return (
        <button
          id={id}
          data-testid={`square-${id}`}
          className={className}
          disabled={disabled}
          style={{ backgroundImage: url }}
        >
          {id}
        </button>
      );
    }
);

jest.mock("../../../hooks/game/useBoard");

describe("BoardUI", () => {
  const mockPlayGame = jest.fn();

  const emptyBoard: GameBoard = [
    Player.Nobody,
    Player.Nobody,
    Player.Nobody,
    Player.Nobody,
    Player.Nobody,
    Player.Nobody,
    Player.Nobody,
    Player.Nobody,
    Player.Nobody,
  ];

  interface Props {
    disabled?: boolean;
    over?: boolean;
    board?: GameBoard;
    winners?: number[];
  }

  const renderBoard = ({
    disabled = false,
    over = false,
    board = emptyBoard,
    winners = [],
  }: Props = {}) => {
    const game = {
      category: {
        id: 5,
        folder: "soccer",
        name: "CATEGORY_SOCCER_PLAYERS",
      },
      userImage: 5,
      computerImage: 6,
      first: Player.User,
      level: PlayingLevel.Easy,
      disabled,
      over,
      result: GameResult.Started,
      winners,
    } as State;

    (useBoard as jest.Mock).mockReturnValue({
      gameBoard: board,
      playGame: mockPlayGame,
    });

    const dispatch = jest.fn();
    const handleGameOver = jest.fn();

    const utils = render(
      <GameContext.Provider value={{ game, dispatch }}>
        <BoardUI onGameOver={handleGameOver} />
      </GameContext.Provider>
    );

    return { ...utils, mockPlayGame };
  };

  it("should render the BoardUI component correctly", () => {
    const { asFragment } = renderBoard();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the wrapper div with class board", () => {
    renderBoard();

    expect(screen.queryByTestId("board")).toHaveClass("board");
  });

  it("should render the board when disabled is true", async () => {
    const user = userEvent.setup();
    renderBoard({ disabled: true });
    const firstSquare = screen.queryByTestId("square-0");

    await user.click(firstSquare!);

    expect(mockPlayGame).toHaveBeenCalled();
  });

  it("should render the board when over is true", async () => {
    const user = userEvent.setup();
    const over = true;
    const board: GameBoard = [
      Player.Computer,
      Player.Computer,
      Player.Computer,
      Player.User,
      Player.User,
      Player.Computer,
      Player.User,
      Player.User,
      Player.Computer,
    ];
    const winners = [0, 1, 2];
    renderBoard({ over, board, winners });
    const firstSquare = screen.queryByTestId("square-0");

    await user.click(firstSquare!);

    expect(mockPlayGame).not.toHaveBeenCalled();
  });

  it("should call the click handler", async () => {
    const user = userEvent.setup();
    const disabled = true;
    renderBoard({ disabled });
    const firstSquare = screen.queryByTestId("square-0");

    await user.click(firstSquare!);

    expect(mockPlayGame).toHaveBeenCalled();
  });
});
