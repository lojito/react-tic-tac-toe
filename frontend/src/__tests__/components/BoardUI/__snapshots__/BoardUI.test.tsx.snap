// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardUI should render the BoardUI component correctly 1`] = `
<DocumentFragment>
  <div
    class="board"
    data-testid="board"
  >
    <button
      class=""
      data-testid="square-0"
      disabled=""
      id="0"
      style="background-image: url(undefined/images/default.jpg);"
    >
      0
    </button>
    <button
      class=""
      data-testid="square-1"
      disabled=""
      id="1"
      style="background-image: url(undefined/images/default.jpg);"
    >
      1
    </button>
    <button
      class=""
      data-testid="square-2"
      disabled=""
      id="2"
      style="background-image: url(undefined/images/default.jpg);"
    >
      2
    </button>
    <button
      class=""
      data-testid="square-3"
      disabled=""
      id="3"
      style="background-image: url(undefined/images/default.jpg);"
    >
      3
    </button>
    <button
      class=""
      data-testid="square-4"
      disabled=""
      id="4"
      style="background-image: url(undefined/images/default.jpg);"
    >
      4
    </button>
    <button
      class=""
      data-testid="square-5"
      disabled=""
      id="5"
      style="background-image: url(undefined/images/default.jpg);"
    >
      5
    </button>
    <button
      class=""
      data-testid="square-6"
      disabled=""
      id="6"
      style="background-image: url(undefined/images/default.jpg);"
    >
      6
    </button>
    <button
      class=""
      data-testid="square-7"
      disabled=""
      id="7"
      style="background-image: url(undefined/images/default.jpg);"
    >
      7
    </button>
    <button
      class=""
      data-testid="square-8"
      disabled=""
      id="8"
      style="background-image: url(undefined/images/default.jpg);"
    >
      8
    </button>
  </div>
</DocumentFragment>
`;
